### DISCUSSION TITLE
How to handle routing?

### DISCUSSION THREAD

**Discussion by RareSecond** [statelyai/xstate#2161] [Q&A]
We are currently looking towards XState to help us manage our flows a lot more clearly. We currently have Redux, but there's so much "magic" there and no clear structure, that it's very hard for a new developer to come in and instantly get acquainted with the flow he's currently looking at.

XState seems like the perfect fit for our product, which is basically a bunch of modules that guide the users through a bunch of flows.

However, while I love the idea behind XState, I'm missing 1 required real-life implementation: how it can easily be integrated with browser routing. I've stumbled across [this comment](https://github.com/davidkpiano/xstate/issues/51#issuecomment-361595766) from @davidkpiano, but the link mention at the top of the post appears to be broken. I have spent some time looking around for other definitive answers, but have found none.

The reason why routing seems so important for me, is giving the user the ability to use the browser's native capabilities of going back.

Let's say I have a simple registration form that has three states:

- Initial, which is a form that asks you for your name
- Profile, which is a form that might ask personal questions (e.g. age, gender, etc..)
- Submit, which will present you with a button to submit the information
- Submitting, triggered by pressing the button, resulting in a loading button
- Success, which will be triggered on a successful API call and will present you with a small success message

Say now that every one of these states should correspond with a history entry in your browser. We might want to make it easier for the user to press the back button when they are entering the personal questions because they notice their name was misspelled, how would we do this?

---------

I started to work on a very small POC to see what I could come up with, but quickly got stuck. This was my very early draft:

```
createMachine({
  ...
  states: {
    ...
    userInformation: {
      entry: () => {
        history.push('/personal');
      },
      on: {
        NEXT: 'submit',
      },
    },
    ...
  },
});
```

Seems pretty straightforward. But as you can see from this small example, this will quickly get pretty verbose. Will we need to add a full `entry` object with the `history.push` action on every state? Is there a solution to have a "global" `entry` action that can be fired with some context about the step that fired it?

**Comment by ChrisShank** [comment]
There have been some ideas being thrown around in the [xstate discord](https://discord.com/channels/795785288994652170/799416911259369472/835527057403281459) recently, but I don't think there has been much consensus about a robust approach yet.

I haven't used it myself, but there is also this project called xstate-router that you could look into: https://github.com/carloslfu/xstate-router

There is **definitely** a need for xstate to more easily interact/sync with routes, but I think its still an a very exploratory stage!

**Comment by tmeindle** [comment]
I used the project xstate-router but found it a little buggy and no mechanism for switching to the correct initial state when a route is navigated directly to.  But, I liked the concept so rewrote it in ES6 and enhanced it a little..  Might be useful to someone: https://codesandbox.io/s/xstate-routed-machines-forked-qlp3l
